File: moonlight-android/app/src/main/java/com/limelight/relay/RelayTcpTunnel.java
/**
 * VipleStream: TCP Tunnel through Relay
 * Creates a local TCP listener that tunnels connections through
 * the relay WebSocket to a target server's TCP port.
 * Used for RTSP handshake when server is behind double NAT.
 *
 * IMPORTANT: moonlight-common-c's RTSP client opens a NEW TCP connection for each
 * request-response exchange (OPTIONS, DESCRIBE, SETUP x3, PLAY). The server closes
 * the TCP connection after sending the response (recv returns 0 = EOF).
 * Therefore this tunnel must accept MULTIPLE sequential TCP connections over the
 * same WebSocket relay channel.
 *
 * Ported from the PC/Qt version (relaytcptunnel.cpp).
 */
package com.limelight.relay;

import android.util.Base64;
import android.util.Log;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.net.SocketTimeoutException;
import java.security.SecureRandom;
import java.security.cert.X509Certificate;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSocket;
import javax.net.ssl.SSLSocketFactory;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;

import org.json.JSONObject;

public class RelayTcpTunnel extends Thread {

    private static final String TAG = "RelayTunnel";

    private final String mRelayUrl;
    private final String mRelayPsk;
    private final String mServerUuid;
    private final int mTargetPort;

    private volatile int mLocalPort = 0;
    private volatile boolean mStop = false;
    private volatile ServerSocket mServerSocket;
    private volatile Socket mRelaySocket;

    public RelayTcpTunnel(String relayUrl, String relayPsk,
                          String serverUuid, int targetPort) {
        super("RelayTcpTunnel");
        setDaemon(true);
        mRelayUrl = relayUrl;
        mRelayPsk = relayPsk;
        mServerUuid = serverUuid;
        mTargetPort = targetPort;
    }

    /**
     * Start the tunnel thread and wait for it to begin listening.
     * @return the local port to connect to, or 0 on failure.
     */
    public int startAndGetPort() {
        start();
        // Wait up to 3 seconds for mLocalPort to be assigned
        for (int i = 0; i < 300 && mLocalPort == 0 && !mStop; i++) {
            try {
                Thread.sleep(10);
            } catch (InterruptedException e) {
                break;
            }
        }
        return mLocalPort;
    }

    /**
     * Stop the tunnel. Closes sockets to unblock threads.
     */
    public void stopTunnel() {
        mStop = true;
        closeQuietly(mServerSocket);
        closeQuietly(mRelaySocket);
    }

    @Override
    public void run() {
        ServerSocket serverSock = null;
        Socket relaySock = null;

        try {
            // Create local TCP server on loopback, random port
            serverSock = new ServerSocket(0, 1, InetAddress.getByName("127.0.0.1"));
            mServerSocket = serverSock;
            mLocalPort = serverSock.getLocalPort();

            Log.i(TAG, "Listening on localhost:" + mLocalPort + " -> server "
                    + mServerUuid.substring(0, Math.min(8, mServerUuid.length()))
                    + " port " + mTargetPort);

            // --- Connect to relay via WebSocket (persistent for entire tunnel lifetime) ---
            boolean useTls;
            String host;
            int port;
            String raw = mRelayUrl;
            if (raw.startsWith("wss://")) {
                useTls = true;
                raw = raw.substring(6);
            } else if (raw.startsWith("ws://")) {
                useTls = false;
                raw = raw.substring(5);
            } else {
                useTls = false;
            }
            int colonIdx = raw.lastIndexOf(':');
            if (colonIdx > 0) {
                host = raw.substring(0, colonIdx);
                try {
                    port = Integer.parseInt(raw.substring(colonIdx + 1));
                } catch (NumberFormatException e) {
                    port = useTls ? 443 : 9999;
                }
            } else {
                host = raw;
                port = useTls ? 443 : 9999;
            }

            if (useTls) {
                SSLSocketFactory factory = createTrustAllFactory();
                relaySock = factory.createSocket();
                relaySock.setSoTimeout(10000);
                relaySock.connect(new java.net.InetSocketAddress(host, port), 10000);
                ((SSLSocket) relaySock).startHandshake();
            } else {
                relaySock = new Socket();
                relaySock.setSoTimeout(10000);
                relaySock.connect(new java.net.InetSocketAddress(host, port), 10000);
            }
            mRelaySocket = relaySock;

            InputStream relayIn = relaySock.getInputStream();
            OutputStream relayOut = relaySock.getOutputStream();

            // WS handshake
            byte[] wsKeyBytes = new byte[16];
            new SecureRandom().nextBytes(wsKeyBytes);
            String wsKey = Base64.encodeToString(wsKeyBytes, Base64.NO_WRAP);

            String httpReq = "GET / HTTP/1.1\r\n" +
                    "Host: " + host + "\r\n" +
                    "Upgrade: websocket\r\n" +
                    "Connection: Upgrade\r\n" +
                    "Sec-WebSocket-Key: " + wsKey + "\r\n" +
                    "Sec-WebSocket-Version: 13\r\n" +
                    "\r\n";
            relayOut.write(httpReq.getBytes("UTF-8"));
            relayOut.flush();

            String hsResp = readUntilHeaderEnd(relayIn, 10000);
            if (hsResp == null || !hsResp.contains("101")) {
                Log.w(TAG, "WS handshake failed");
                return;
            }

            // Register with relay
            String tunnelUuid = "tunnel-"
                    + mServerUuid.substring(0, Math.min(8, mServerUuid.length()))
                    + "-" + randomHex(6);

            String pskHash = computePskHash(mRelayPsk, tunnelUuid);

            JSONObject regMsg = new JSONObject();
            regMsg.put("type", "register");
            regMsg.put("uuid", tunnelUuid);
            regMsg.put("role", "client");
            regMsg.put("psk_hash", pskHash);
            relayOut.write(wsFrameMasked(regMsg.toString().getBytes("UTF-8")));
            relayOut.flush();

            // Consume register response
            relaySock.setSoTimeout(5000);
            wsRead(relayIn);

            // Request TCP tunnel to server (once, for the whole session)
            JSONObject tunnelReq = new JSONObject();
            tunnelReq.put("type", "tcp_tunnel_open");
            tunnelReq.put("target_uuid", mServerUuid);
            tunnelReq.put("target_port", mTargetPort);
            tunnelReq.put("tunnel_id", tunnelUuid);
            relayOut.write(wsFrameMasked(tunnelReq.toString().getBytes("UTF-8")));
            relayOut.flush();

            Log.i(TAG, "TCP tunnel registered (uuid=" + tunnelUuid + "), accepting connections");

            // ===================================================================
            // Main loop: accept multiple sequential RTSP connections
            // ===================================================================
            int connectionCount = 0;

            while (!mStop) {
                // Wait for an incoming connection (short timeout so we can check mStop)
                serverSock.setSoTimeout(1000);
                Socket localClient;
                try {
                    localClient = serverSock.accept();
                } catch (SocketTimeoutException e) {
                    // Drain any WS messages while waiting (keep-alive pings, etc.)
                    drainRelayMessages(relaySock, relayIn, relayOut);
                    continue;
                }

                connectionCount++;
                Log.i(TAG, "Connection #" + connectionCount + " from RTSP client");

                // Signal server to open a new TCP connection (for connections after the first)
                if (connectionCount > 1) {
                    JSONObject reconnReq = new JSONObject();
                    reconnReq.put("type", "tcp_tunnel_reconnect");
                    reconnReq.put("target_uuid", mServerUuid);
                    reconnReq.put("target_port", mTargetPort);
                    reconnReq.put("tunnel_id", tunnelUuid);
                    relayOut.write(wsFrameMasked(reconnReq.toString().getBytes("UTF-8")));
                    relayOut.flush();
                    Log.i(TAG, "Sent reconnect request for connection #" + connectionCount);
                }

                // Relay data for this connection until it closes
                InputStream localIn = localClient.getInputStream();
                OutputStream localOut = localClient.getOutputStream();
                boolean tunnelClosed = false;
                byte[] localBuf = new byte[65536];

                while (!mStop && !tunnelClosed && !localClient.isClosed()) {
                    boolean activity = false;

                    // Local -> Relay: read from local client (short timeout)
                    try {
                        localClient.setSoTimeout(50);
                        int n = localIn.read(localBuf);
                        if (n > 0) {
                            String encoded = Base64.encodeToString(localBuf, 0, n, Base64.NO_WRAP);
                            JSONObject dataMsg = new JSONObject();
                            dataMsg.put("type", "tcp_tunnel_data");
                            dataMsg.put("tunnel_id", tunnelUuid);
                            dataMsg.put("target_uuid", mServerUuid);
                            dataMsg.put("data", encoded);
                            relayOut.write(wsFrameMasked(dataMsg.toString().getBytes("UTF-8")));
                            relayOut.flush();
                            activity = true;
                            Log.i(TAG, "#" + connectionCount + " Local->Relay: " + n + " bytes");
                        } else if (n < 0) {
                            // Local client disconnected
                            Log.i(TAG, "#" + connectionCount + " Local client disconnected");
                            break;
                        }
                    } catch (SocketTimeoutException e) {
                        // No data from local client yet, that's fine
                    }

                    // Relay -> Local: read WS frames from relay
                    try {
                        relaySock.setSoTimeout(100);
                        String wsData = wsRead(relayIn);
                        if (wsData != null && !wsData.isEmpty()) {
                            JSONObject obj = new JSONObject(wsData);
                            String msgType = obj.optString("type", "");

                            if ("tcp_tunnel_data".equals(msgType)) {
                                byte[] decoded = Base64.decode(
                                        obj.optString("data", ""), Base64.NO_WRAP);
                                localOut.write(decoded);
                                localOut.flush();
                                activity = true;
                                Log.i(TAG, "#" + connectionCount + " Relay->Local: "
                                        + decoded.length + " bytes");

                            } else if ("tcp_tunnel_closed".equals(msgType)) {
                                // Server closed TCP -- normal for per-request RTSP
                                Log.i(TAG, "#" + connectionCount
                                        + " Server closed TCP (normal per-request close)");
                                tunnelClosed = true;

                            } else if ("ping".equals(msgType)) {
                                JSONObject pong = new JSONObject();
                                pong.put("type", "pong");
                                relayOut.write(wsFrameMasked(
                                        pong.toString().getBytes("UTF-8")));
                                relayOut.flush();
                            } else {
                                Log.i(TAG, "#" + connectionCount + " Got msg: "
                                        + msgType + " (ignored)");
                            }
                        }
                    } catch (SocketTimeoutException e) {
                        // No data from relay yet
                    }

                    if (!activity && !tunnelClosed) {
                        try {
                            Thread.sleep(5);
                        } catch (InterruptedException e) {
                            break;
                        }
                    }
                }

                // Close local client socket
                closeQuietly(localClient);
                Log.i(TAG, "Connection #" + connectionCount + " completed");
            }

            Log.i(TAG, "Tunnel closed after " + connectionCount + " connections");

        } catch (Exception e) {
            if (!mStop) {
                Log.w(TAG, "Tunnel error: " + e.getMessage());
            }
        } finally {
            closeQuietly(relaySock);
            closeQuietly(serverSock);
        }
    }

    // -----------------------------------------------------------------------
    // WebSocket frame helpers (duplicated from RelayClient for thread safety)
    // -----------------------------------------------------------------------

    private static byte[] wsFrameMasked(byte[] payload) {
        int len = payload.length;
        int headerLen;
        byte[] frame;

        if (len < 126) {
            headerLen = 2 + 4;
            frame = new byte[headerLen + len];
            frame[0] = (byte) 0x81;
            frame[1] = (byte) (0x80 | len);
        } else if (len < 65536) {
            headerLen = 2 + 2 + 4;
            frame = new byte[headerLen + len];
            frame[0] = (byte) 0x81;
            frame[1] = (byte) (0x80 | 126);
            frame[2] = (byte) (len >> 8);
            frame[3] = (byte) (len & 0xFF);
        } else {
            headerLen = 2 + 8 + 4;
            frame = new byte[headerLen + len];
            frame[0] = (byte) 0x81;
            frame[1] = (byte) (0x80 | 127);
            for (int i = 7; i >= 0; i--) {
                frame[2 + (7 - i)] = (byte) ((len >> (i * 8)) & 0xFF);
            }
        }

        byte[] mask = new byte[4];
        new SecureRandom().nextBytes(mask);
        int maskOffset = headerLen - 4;
        System.arraycopy(mask, 0, frame, maskOffset, 4);

        for (int i = 0; i < len; i++) {
            frame[headerLen + i] = (byte) (payload[i] ^ mask[i % 4]);
        }

        return frame;
    }

    private static String wsRead(InputStream in) {
        try {
            byte[] hdr = readExact(in, 2);
            if (hdr == null) return null;

            int len = hdr[1] & 0x7F;
            if (len == 126) {
                byte[] ext = readExact(in, 2);
                if (ext == null) return null;
                len = ((ext[0] & 0xFF) << 8) | (ext[1] & 0xFF);
            } else if (len == 127) {
                byte[] ext = readExact(in, 8);
                if (ext == null) return null;
                len = 0;
                for (int i = 0; i < 8; i++) {
                    len = (len << 8) | (ext[i] & 0xFF);
                }
            }

            if (len <= 0 || len > 1048576) return null;

            byte[] data = readExact(in, len);
            if (data == null) return null;

            return new String(data, "UTF-8");
        } catch (Exception e) {
            return null;
        }
    }

    private static byte[] readExact(InputStream in, int n) throws IOException {
        byte[] buf = new byte[n];
        int off = 0;
        while (off < n) {
            int r = in.read(buf, off, n - off);
            if (r < 0) return null;
            off += r;
        }
        return buf;
    }

    // -----------------------------------------------------------------------
    // Internal helpers
    // -----------------------------------------------------------------------

    private static String readUntilHeaderEnd(InputStream in, int timeoutMs) {
        try {
            StringBuilder sb = new StringBuilder();
            long deadline = System.currentTimeMillis() + timeoutMs;
            while (sb.length() < 8192) {
                if (System.currentTimeMillis() > deadline) return null;
                int b = in.read();
                if (b < 0) return null;
                sb.append((char) b);
                if (sb.length() >= 4) {
                    String tail = sb.substring(sb.length() - 4);
                    if ("\r\n\r\n".equals(tail)) {
                        return sb.toString();
                    }
                }
            }
        } catch (IOException e) {
            // fall through
        }
        return null;
    }

    private static String computePskHash(String psk, String uuid) {
        if (psk == null || psk.isEmpty()) return "";
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            SecretKeySpec keySpec = new SecretKeySpec(psk.getBytes("UTF-8"), "HmacSHA256");
            mac.init(keySpec);
            byte[] hmacResult = mac.doFinal(uuid.getBytes("UTF-8"));
            return bytesToHex(hmacResult).substring(0, 16);
        } catch (Exception e) {
            Log.w(TAG, "HMAC failed: " + e.getMessage());
            return "";
        }
    }

    private static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder(bytes.length * 2);
        for (byte b : bytes) {
            sb.append(String.format("%02x", b & 0xFF));
        }
        return sb.toString();
    }

    private static String randomHex(int chars) {
        SecureRandom rng = new SecureRandom();
        byte[] bytes = new byte[(chars + 1) / 2];
        rng.nextBytes(bytes);
        return bytesToHex(bytes).substring(0, chars);
    }

    /**
     * Drain any pending WS messages from the relay while waiting for accept().
     * Responds to pings to keep the connection alive.
     */
    private void drainRelayMessages(Socket relaySock, InputStream relayIn,
                                    OutputStream relayOut) {
        try {
            relaySock.setSoTimeout(50);
            String wsData = wsRead(relayIn);
            if (wsData != null && !wsData.isEmpty()) {
                JSONObject obj = new JSONObject(wsData);
                String msgType = obj.optString("type", "");
                if ("ping".equals(msgType)) {
                    JSONObject pong = new JSONObject();
                    pong.put("type", "pong");
                    relayOut.write(wsFrameMasked(pong.toString().getBytes("UTF-8")));
                    relayOut.flush();
                }
            }
        } catch (SocketTimeoutException e) {
            // Expected when no data available
        } catch (Exception e) {
            // Ignore drain errors
        }
    }

    private static SSLSocketFactory createTrustAllFactory() {
        try {
            TrustManager[] trustAll = new TrustManager[]{
                    new X509TrustManager() {
                        @Override
                        public void checkClientTrusted(X509Certificate[] chain, String authType) {}

                        @Override
                        public void checkServerTrusted(X509Certificate[] chain, String authType) {}

                        @Override
                        public X509Certificate[] getAcceptedIssuers() {
                            return new X509Certificate[0];
                        }
                    }
            };
            SSLContext ctx = SSLContext.getInstance("TLS");
            ctx.init(null, trustAll, new SecureRandom());
            return ctx.getSocketFactory();
        } catch (Exception e) {
            throw new RuntimeException("Failed to create TrustAll SSLContext", e);
        }
    }

    private static void closeQuietly(Socket sock) {
        if (sock != null) {
            try { sock.close(); } catch (IOException ignored) {}
        }
    }

    private static void closeQuietly(ServerSocket sock) {
        if (sock != null) {
            try { sock.close(); } catch (IOException ignored) {}
        }
    }
}

File: moonlight-qt/app/backend/relaytcptunnel.cpp
/**
 * VipleStream: TCP Tunnel through Relay
 * Tunnels local TCP connections through the relay WebSocket to a remote server.
 *
 * IMPORTANT: moonlight-common-c's RTSP client opens a NEW TCP connection for each
 * request-response exchange (OPTIONS, DESCRIBE, SETUP×3, PLAY). The server closes
 * the TCP connection after sending the response (recv returns 0 = EOF).
 * Therefore this tunnel must accept MULTIPLE sequential TCP connections over the
 * same WebSocket relay channel.
 *
 * Protocol: sends "tcp_tunnel_data" messages with base64-encoded TCP data.
 * Per-request flow:
 *   1. RTSP client connects to our local port
 *   2. Client sends request → we forward as tcp_tunnel_data
 *   3. Sunshine relays back response as tcp_tunnel_data
 *   4. Sunshine sends tcp_tunnel_closed (RTSP server closed TCP)
 *   5. We close local client socket (EOF for RTSP client)
 *   6. Go back to step 1 for the next RTSP request
 */

#include "relaytcptunnel.h"
#include "relaylookup.h"

#include <QUrl>
#include <QJsonDocument>
#include <QJsonObject>
#include <QCryptographicHash>
#include <QMessageAuthenticationCode>
#include <QRandomGenerator>
#include <SDL_log.h>

#ifdef _WIN32
#include <winsock2.h>
#include <ws2tcpip.h>
#else
#include <sys/socket.h>
#include <poll.h>
#include <unistd.h>
#endif

// Reuse WS frame helpers (client→server must be masked per RFC 6455)
static QByteArray tunWsFrame(const QByteArray &payload) {
    QByteArray frame;
    frame.append((char)0x81);
    int len = payload.size();
    if (len < 126) {
        frame.append((char)(0x80 | len));
    } else if (len < 65536) {
        frame.append((char)(0x80 | 126));
        frame.append((char)(len >> 8));
        frame.append((char)(len & 0xFF));
    } else {
        frame.append((char)(0x80 | 127));
        for (int i = 7; i >= 0; i--) frame.append((char)((len >> (i * 8)) & 0xFF));
    }
    quint32 maskVal = QRandomGenerator::global()->generate();
    QByteArray mask((const char *)&maskVal, 4);
    frame.append(mask);
    for (int i = 0; i < len; i++) {
        frame.append(payload[i] ^ mask[i % 4]);
    }
    return frame;
}

static QByteArray tunReadExact(QAbstractSocket *sock, int n, int timeoutMs) {
    QByteArray buf;
    while (buf.size() < n) {
        if (sock->bytesAvailable() == 0 && !sock->waitForReadyRead(timeoutMs)) break;
        buf.append(sock->read(n - buf.size()));
    }
    return buf;
}

static QByteArray tunWsRead(QAbstractSocket *sock, int timeoutMs) {
    QByteArray hdr = tunReadExact(sock, 2, timeoutMs);
    if (hdr.size() < 2) return QByteArray();
    int len = (quint8)hdr[1] & 0x7F;
    if (len == 126) {
        QByteArray ext = tunReadExact(sock, 2, timeoutMs);
        if (ext.size() < 2) return QByteArray();
        len = ((quint8)ext[0] << 8) | (quint8)ext[1];
    } else if (len == 127) {
        QByteArray ext = tunReadExact(sock, 8, timeoutMs);
        len = 0;
        for (int i = 0; i < 8; i++) len = (len << 8) | (quint8)ext[i];
    }
    if (len == 0 || len > 1048576) return QByteArray();
    return tunReadExact(sock, len, timeoutMs);
}

RelayTcpTunnel::RelayTcpTunnel(const QString &relayUrl, const QString &relayPsk,
                                 const QString &serverUuid, uint16_t targetPort)
    : m_RelayUrl(relayUrl), m_RelayPsk(relayPsk),
      m_ServerUuid(serverUuid), m_TargetPort(targetPort) {}

RelayTcpTunnel::~RelayTcpTunnel() {
    stop();
    wait(10000);
}

uint16_t RelayTcpTunnel::startAndGetPort() {
    QThread::start();
    // Wait for the port to be assigned by the thread
    for (int i = 0; i < 300 && m_LocalPort == 0 && !m_Stop; i++) {
        QThread::msleep(10);
    }
    return m_LocalPort;
}

void RelayTcpTunnel::stop() {
    m_Stop = true;
}

void RelayTcpTunnel::run() {
    // Create TCP server INSIDE the thread (Qt objects must be used in their own thread)
    QTcpServer server;
    server.listen(QHostAddress::LocalHost, 0);
    m_LocalPort = server.serverPort();

    SDL_LogInfo(SDL_LOG_CATEGORY_APPLICATION,
                "[VIPLE-TUNNEL] Listening on localhost:%d -> server %s port %d",
                m_LocalPort.load(), qPrintable(m_ServerUuid.left(8)), m_TargetPort);

    // --- Connect to relay via WebSocket (persistent for entire tunnel lifetime) ---
    QUrl url(m_RelayUrl);
    bool useTls = (url.scheme() == "wss");
    QString host = url.host();
    int port = url.port(useTls ? 443 : 9999);
    if (host.isEmpty()) {
        QString raw = m_RelayUrl;
        if (raw.startsWith("wss://")) { raw = raw.mid(6); useTls = true; }
        else if (raw.startsWith("ws://")) { raw = raw.mid(5); }
        int ci = raw.lastIndexOf(':');
        if (ci > 0) { host = raw.left(ci); port = raw.mid(ci + 1).toInt(); }
        else host = raw;
    }

    QAbstractSocket *relaySock;
    if (useTls) {
        QSslSocket *ssl = new QSslSocket();
        ssl->setPeerVerifyMode(QSslSocket::VerifyNone);
        ssl->connectToHostEncrypted(host, port);
        if (!ssl->waitForEncrypted(10000)) {
            SDL_LogWarn(SDL_LOG_CATEGORY_APPLICATION, "[VIPLE-TUNNEL] TLS connect failed");
            delete ssl; return;
        }
        relaySock = ssl;
    } else {
        QTcpSocket *tcp = new QTcpSocket();
        tcp->connectToHost(host, port);
        if (!tcp->waitForConnected(10000)) {
            delete tcp; return;
        }
        relaySock = tcp;
    }

    // WS handshake
    QByteArray wsKey = QByteArray(16, 0);
    QRandomGenerator::global()->fillRange((quint32 *)wsKey.data(), 4);
    QString httpReq = QString("GET / HTTP/1.1\r\nHost: %1\r\nUpgrade: websocket\r\nConnection: Upgrade\r\nSec-WebSocket-Key: %2\r\nSec-WebSocket-Version: 13\r\n\r\n")
        .arg(host).arg(QString(wsKey.toBase64()));
    relaySock->write(httpReq.toUtf8());
    relaySock->waitForBytesWritten(5000);

    QByteArray hsResp;
    while (hsResp.size() < 8192) {
        if (relaySock->bytesAvailable() == 0 && !relaySock->waitForReadyRead(5000)) break;
        QByteArray chunk = relaySock->peek(relaySock->bytesAvailable());
        int endIdx = chunk.indexOf("\r\n\r\n");
        if (endIdx >= 0) { hsResp.append(relaySock->read(endIdx + 4)); break; }
        else hsResp.append(relaySock->readAll());
    }
    if (!hsResp.contains("101")) {
        SDL_LogWarn(SDL_LOG_CATEGORY_APPLICATION, "[VIPLE-TUNNEL] WS handshake failed");
        delete relaySock; return;
    }

    // Register with relay
    QString tunnelUuid = "tunnel-" + m_ServerUuid.left(8) + "-" +
                         QString::number(QRandomGenerator::global()->generate(), 16).left(6);
    QString pskHash;
    if (!m_RelayPsk.isEmpty()) {
        QMessageAuthenticationCode mac(QCryptographicHash::Sha256);
        mac.setKey(m_RelayPsk.toUtf8());
        mac.addData(tunnelUuid.toUtf8());
        pskHash = mac.result().toHex().left(16);
    }
    QJsonObject regMsg;
    regMsg["type"] = "register";
    regMsg["uuid"] = tunnelUuid;
    regMsg["role"] = "client";
    regMsg["psk_hash"] = pskHash;
    relaySock->write(tunWsFrame(QJsonDocument(regMsg).toJson(QJsonDocument::Compact)));
    relaySock->waitForBytesWritten(5000);
    tunWsRead(relaySock, 5000); // consume register response

    // Request TCP tunnel to server (once, for the whole session)
    QJsonObject tunnelReq;
    tunnelReq["type"] = "tcp_tunnel_open";
    tunnelReq["target_uuid"] = m_ServerUuid;
    tunnelReq["target_port"] = m_TargetPort;
    tunnelReq["tunnel_id"] = tunnelUuid;
    relaySock->write(tunWsFrame(QJsonDocument(tunnelReq).toJson(QJsonDocument::Compact)));
    relaySock->waitForBytesWritten(5000);

    SDL_LogInfo(SDL_LOG_CATEGORY_APPLICATION,
                "[VIPLE-TUNNEL] TCP tunnel registered (uuid=%s), accepting connections",
                qPrintable(tunnelUuid));

    // ===================================================================
    // Main loop: accept multiple sequential RTSP connections
    // Each RTSP request is a new TCP connection (connect → send → recv → close)
    // ===================================================================
    int connectionCount = 0;
    while (!m_Stop) {
        // Wait for an incoming connection (short timeout so we can check m_Stop)
        if (!server.waitForNewConnection(1000)) {
            // Drain any WS messages while waiting (keep-alive, endpoint updates, etc.)
            QByteArray wsData = tunWsRead(relaySock, 50);
            if (!wsData.isEmpty()) {
                QJsonDocument doc = QJsonDocument::fromJson(wsData);
                QString msgType = doc.object()["type"].toString();
                if (msgType == "ping") {
                    // Respond to keep-alive
                    QJsonObject pong;
                    pong["type"] = "pong";
                    relaySock->write(tunWsFrame(QJsonDocument(pong).toJson(QJsonDocument::Compact)));
                    relaySock->waitForBytesWritten(1000);
                }
            }
            continue;
        }

        connectionCount++;
        QTcpSocket *localClient = server.nextPendingConnection();
        SDL_LogInfo(SDL_LOG_CATEGORY_APPLICATION,
                    "[VIPLE-TUNNEL] Connection #%d from RTSP client", connectionCount);

        // Signal Sunshine to open a new TCP connection to RTSP server
        // (for connections after the first one)
        if (connectionCount > 1) {
            QJsonObject reopenReq;
            reopenReq["type"] = "tcp_tunnel_reconnect";
            reopenReq["target_uuid"] = m_ServerUuid;
            reopenReq["target_port"] = m_TargetPort;
            reopenReq["tunnel_id"] = tunnelUuid;
            relaySock->write(tunWsFrame(QJsonDocument(reopenReq).toJson(QJsonDocument::Compact)));
            relaySock->waitForBytesWritten(5000);
            SDL_LogInfo(SDL_LOG_CATEGORY_APPLICATION,
                        "[VIPLE-TUNNEL] Sent reconnect request for connection #%d", connectionCount);
        }

        // Relay data for this connection until it closes
        // Use native socket descriptor for reliable I/O in non-event-loop thread
        qintptr localFd = localClient->socketDescriptor();
        bool tunnelClosed = false;
        while (!m_Stop && !tunnelClosed && localFd != -1 &&
               localClient->state() == QAbstractSocket::ConnectedState) {
            bool activity = false;

            // Local → Relay: poll local socket for incoming data (50ms)
            {
#ifdef _WIN32
                fd_set readSet;
                FD_ZERO(&readSet);
                FD_SET((SOCKET)localFd, &readSet);
                struct timeval tv = {0, 50000}; // 50ms
                int sel = ::select(0, &readSet, nullptr, nullptr, &tv);
#else
                struct pollfd pfd = { (int)localFd, POLLIN, 0 };
                int sel = ::poll(&pfd, 1, 50);
#endif
                if (sel > 0) {
                    char buf[65536];
#ifdef _WIN32
                    int n = ::recv((SOCKET)localFd, buf, sizeof(buf), 0);
#else
                    int n = ::recv((int)localFd, buf, sizeof(buf), 0);
#endif
                    if (n > 0) {
                        QJsonObject dataMsg;
                        dataMsg["type"] = "tcp_tunnel_data";
                        dataMsg["tunnel_id"] = tunnelUuid;
                        dataMsg["target_uuid"] = m_ServerUuid;
                        dataMsg["data"] = QString(QByteArray(buf, n).toBase64());
                        QByteArray frame = tunWsFrame(QJsonDocument(dataMsg).toJson(QJsonDocument::Compact));
                        relaySock->write(frame);
                        relaySock->waitForBytesWritten(5000);
                        activity = true;
                        SDL_LogInfo(SDL_LOG_CATEGORY_APPLICATION,
                                    "[VIPLE-TUNNEL] #%d Local→Relay: %d bytes", connectionCount, n);
                    } else if (n == 0) {
                        // Client disconnected
                        SDL_LogInfo(SDL_LOG_CATEGORY_APPLICATION,
                                    "[VIPLE-TUNNEL] #%d Local client disconnected", connectionCount);
                        break;
                    }
                }
            }

            // Relay → Local: read WS frames, write to local client
            {
                QByteArray wsData = tunWsRead(relaySock, 100);
                if (!wsData.isEmpty()) {
                    QJsonDocument doc = QJsonDocument::fromJson(wsData);
                    QJsonObject obj = doc.object();
                    QString msgType = obj["type"].toString();
                    if (msgType == "tcp_tunnel_data") {
                        QByteArray decoded = QByteArray::fromBase64(obj["data"].toString().toUtf8());
                        // Write directly to socket fd for reliability
#ifdef _WIN32
                        ::send((SOCKET)localFd, decoded.constData(), decoded.size(), 0);
#else
                        ::send((int)localFd, decoded.constData(), decoded.size(), 0);
#endif
                        activity = true;
                        SDL_LogInfo(SDL_LOG_CATEGORY_APPLICATION,
                                    "[VIPLE-TUNNEL] #%d Relay→Local: %d bytes", connectionCount, decoded.size());
                    } else if (msgType == "tcp_tunnel_closed") {
                        // Server-side RTSP closed TCP — this is NORMAL for per-request connections
                        // Close local socket so RTSP client sees EOF (recv returns 0)
                        SDL_LogInfo(SDL_LOG_CATEGORY_APPLICATION,
                                    "[VIPLE-TUNNEL] #%d Server closed TCP (normal per-request close)",
                                    connectionCount);
                        localClient->disconnectFromHost();
                        tunnelClosed = true;
                    } else if (msgType == "ping") {
                        QJsonObject pong;
                        pong["type"] = "pong";
                        relaySock->write(tunWsFrame(QJsonDocument(pong).toJson(QJsonDocument::Compact)));
                        relaySock->waitForBytesWritten(1000);
                    } else {
                        SDL_LogInfo(SDL_LOG_CATEGORY_APPLICATION,
                                    "[VIPLE-TUNNEL] #%d Got msg: %s (ignored)", connectionCount,
                                    qPrintable(msgType));
                    }
                }
            }

            if (!activity && !tunnelClosed) {
                QThread::msleep(5);
            }
        }

        // Wait for local client to fully close
        if (localClient->state() != QAbstractSocket::UnconnectedState) {
            localClient->waitForDisconnected(2000);
        }
        delete localClient;
        SDL_LogInfo(SDL_LOG_CATEGORY_APPLICATION,
                    "[VIPLE-TUNNEL] Connection #%d completed", connectionCount);
    }

    SDL_LogInfo(SDL_LOG_CATEGORY_APPLICATION,
                "[VIPLE-TUNNEL] Tunnel closed after %d connections", connectionCount);
    delete relaySock;
}
